server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -665,7 +671,47 @@ public void onResponse(Releasable response) {
                     joinRequest,
                     ActionListener.runBefore(joinListener, () -> Releasables.close(response))
-                        .delegateFailure((l, ignored) -> processJoinRequest(joinRequest, l))
+                        .delegateFailure((l, ignored) -> processJoinRequest(joinRequest, l.delegateResponse((ignoredListener, e) -> {
+
+                            /*
+                                This prevents a corner case, explained in #ES-11449, occurring as follows:
+                                - Master M is in term T and has cluster state (T, V).
+                                - Node N tries to join the cluster.
+                                - M proposes cluster state (T, V+1) with N in the cluster.
+                                - M accepts its own proposal and commits it to disk.
+                                - M receives no responses. M doesn't know whether the state was accepted by a majority of nodes,
+                                    rejected, or did not reach any nodes.
+                                - There is a re-election and M wins. M publishes cluster state (T+1, V+2).
+                                  Since it's built from the cluster state on disk, N is still in the cluster.
+                                - Since (T, V+1) failed, a FailedToCommitClusterStateException is thrown and N's connection is dropped,
+                                    even though its inclusion in the cluster may have been committed on a majority of master nodes.
+                                - It can rejoin, but this throws a WARN log since it did not restart.
+
+                                The above situation occurs here when a FailedToCommitClusterStateException is thrown.
+                                When we catch this exception, we keep the connection open until the next cluster state update.
+                                N is accepted -> By waiting, we did not close the connection to N unnecessarily
+                                N is rejected -> A new cluster state is published without N in it. Closing is correct here.
+                             */
+                            if (e instanceof FailedToCommitClusterStateException) {
+                                ClusterStateListener clusterStateListener = new ClusterStateListener() {
+                                    @Override
+                                    public void clusterChanged(ClusterChangedEvent event) {
+                                        // Keep the connection open until the next committed state
+                                        if (event.state().nodes().getMasterNode() != null) {
+                                            Releasables.close(response);
+                                            // Remove this listener to avoid memory leaks
+                                            clusterService.removeListener(this);
+                                            joinListener.onResponse(null);
```

Comment:
Maybe we should check to see if the joining node is in the new cluster state, and fail the listener with `e` if not.